New module: shutdown

aminvakil opened the pull request:
##### SUMMARY
New module: shutdown

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/modules/system/shutdown.py

##### ADDITIONAL INFORMATION
Firstly brought up in https://github.com/ansible/ansible/pull/70936, but as ansible team is not "accepting new modules to ansible/ansible directly", here it is.

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+    to specify locations to search if the default paths do not work.
+description:
+    - Shut downs a machine.
+version_added: "2.11"
```
```suggestion
version_added: "1.0.0"
```
(or 1.1.0 if it isn't merged until Friday morning)

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+module: shutdown
+short_description: Shut down a machine
+notes:
+  - C(PATH) is ignored on the remote node when searching for the C(shutdown) command. Use C(search_paths)
```
```suggestion
  - C(PATH) is ignored on the remote node when searching for the C(shutdown) command. Use I(search_paths)
```

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+    description:
+      - Paths to search on the remote machine for the C(shutdown) command.
+      - I(Only) these paths will be searched for the C(shutdown) command. C(PATH) is ignored in the remote node when searching for the C(shutdown) command.
+    type: list
```
```suggestion
    type: list
    elements: path
```

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+
+EXAMPLES = r'''
+- name: Unconditionally shut down the machine with all defaults
+  shutdown:
```
```suggestion
  community.general.shutdown:
```

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+    pass
+
+
+class ActionModule(ActionBase):
```
This is an action module, and **not** a module.

The action plugin needs to move to `plugins/action/system/` (and a symlink from `plugins/action/` to that file). The module file needs to stay in `plugins/modules/system/` containing the documentation.

Andersson007 commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+    - Shut downs a machine.
+version_added: "2.11"
+options:
+  pre_shutdown_delay:
```
```suggestion
  delay:
```
maybe just `delay`? There can't be something post, i think, so, people will be happy to type less

Andersson007 commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,293 @@
+      - Message to display to users before shutdown.
+    type: str
+    default: Shut down initiated by Ansible
+
```
```suggestion
```

Andersson007 commented:
i'd also put:
```
seealso:
- module: ansible.builtin.reboot
```

aminvakil commented:
@felixfontein @Andersson007 Thanks for your suggestions, all applied.
It still has got a problem which I'm working on it, it shuts down the vm correctly, but ansible fails afterwards as it cannot connect to the remote server with this error:
```
fatal: [192.168.46.3]: FAILED! => {
    "msg": "Failed to connect to the host via ssh: OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly"
}
```
The remote module doesn't have this problem because it runs a `validate` function which keeps checking the remote server to connect to it again after it got reboot, I've removed the function as it doesn't apply here.
Also running with -vvvvv gives me this:
```
TASK [run the new module] ***********************************************************************************************************************************************************************************************************
task path: /home/user/dev/ansible/collections/ansible_collections/community/general/testmod.yml:4
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 192.168.46.3 '/bin/sh -c '"'"'echo ~root && sleep 0'"'"''
<192.168.46.3> (0, b'/root\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 192.168.46.3 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791 `" && echo ansible-tmp-1596122604.463402-13282-199918880950791="` echo /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791 `" ) && sleep 0'"'"''
<192.168.46.3> (0, b'ansible-tmp-1596122604.463402-13282-199918880950791=/root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
Using module file /home/user/github/ansible/lib/ansible/modules/setup.py
<192.168.46.3> PUT /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmp7f40d7x7 TO /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_setup.py
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: PlayContext set sftp_extra_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 '[192.168.46.3]'
<192.168.46.3> (0, b'sftp> put /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmp7f40d7x7 /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_setup.py\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug2: Remote version: 3\r\ndebug2: Server supports extension "posix-rename@openssh.com" revision 1\r\ndebug2: Server supports extension "statvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "fstatvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "hardlink@openssh.com" revision 1\r\ndebug2: Server supports extension "fsync@openssh.com" revision 1\r\ndebug3: Sent message fd 3 T:16 I:1\r\ndebug3: SSH_FXP_REALPATH . -> /root size 0\r\ndebug3: Looking up /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmp7f40d7x7\r\ndebug3: Sent message fd 3 T:17 I:2\r\ndebug3: Received stat reply T:101 I:2\r\ndebug1: Couldn\'t stat remote file: No such file or directory\r\ndebug3: Sent message SSH2_FXP_OPEN I:3 P:/root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_setup.py\r\ndebug3: Sent message SSH2_FXP_WRITE I:4 O:0 S:32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 4 32768 bytes at 0\r\ndebug3: Sent message SSH2_FXP_WRITE I:5 O:32768 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:6 O:65536 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:7 O:98304 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:8 O:131072 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:9 O:163840 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:10 O:196608 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:11 O:229376 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:12 O:262144 S:5973\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 5 32768 bytes at 32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 6 32768 bytes at 65536\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 7 32768 bytes at 98304\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 8 32768 bytes at 131072\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 9 32768 bytes at 163840\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 10 32768 bytes at 196608\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 11 32768 bytes at 229376\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 12 5973 bytes at 262144\r\ndebug3: Sent message SSH2_FXP_CLOSE I:4\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 192.168.46.3 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/ /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_setup.py && sleep 0'"'"''
<192.168.46.3> (0, b'', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 -tt 192.168.46.3 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_setup.py && sleep 0'"'"''
<192.168.46.3> (0, b'\r\n{"invocation": {"module_args": {"filter": "*", "gather_subset": ["min"], "fact_path": "/etc/ansible/facts.d", "gather_timeout": 10}}, "ansible_facts": {"ansible_distribution_file_parsed": true, "ansible_fips": false, "ansible_service_mgr": "systemd", "ansible_user_id": "root", "module_setup": true, "ansible_ssh_host_key_rsa_public": "AAAAB3NzaC1yc2EAAAADAQABAAABAQDKVwXJ6N3HYf98ZJekoJISgcWxv3wCtgObADzbfUbTYH0D2rlpKMvBW0zjS6Eg1YydeGfIwsvwcunwGcAzGoopmMsRHx3VSeGccz2wnvVi5/mufnkesjQyr49FjtqQKvQbHSyTXRbrmwdWURrUSAXxgVOZ5RBdQag8ahgymhr0eT5q7CDwTZEBbVAKoHBGoWh2mFm3kovB0kj1dBYbPWTQo+3ArPB5h0A6rUjW8WMAf9uA3PI2xVoXKWC0kEwfTG3xlEPDxrgxyxfJ8CAsFXnqTpzSv9esZe1KUQQmgo4soHB2fbMf9bw/Z6ASmpZh+xYKjASnn2KMjzYp+QZwUvPD", "gather_subset": ["min"], "ansible_architecture": "x86_64", "ansible_kernel_version": "#1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07)", "ansible_distribution_version": "10", "ansible_domain": "", "ansible_distribution_file_path": "/etc/os-release", "ansible_date_time": {"weekday_number": "4", "iso8601_basic_short": "20200730T152324", "tz": "UTC", "weeknumber": "30", "hour": "15", "year": "2020", "minute": "23", "tz_offset": "+0000", "month": "07", "epoch": "1596122604", "iso8601_micro": "2020-07-30T15:23:24.060399Z", "weekday": "Thursday", "time": "15:23:24", "date": "2020-07-30", "iso8601": "2020-07-30T15:23:24Z", "tz_dst": "UTC", "day": "30", "iso8601_basic": "20200730T152324060348", "second": "24"}, "ansible_ssh_host_key_ed25519_public": "AAAAC3NzaC1lZDI1NTE5AAAAIHurpETbUG5ipy1Y6jK1VzC3ExFhhYrRZSLCHrdUNgPD", "ansible_distribution_file_variety": "Debian", "ansible_env": {"LANG": "en_US.UTF-8", "TERM": "xterm-256color", "SHELL": "/bin/bash", "XDG_RUNTIME_DIR": "/run/user/0", "SHLVL": "0", "SSH_TTY": "/dev/pts/0", "PWD": "/root", "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "XDG_SESSION_TYPE": "tty", "SSH_CLIENT": "192.168.51.100 36208 22", "LOGNAME": "root", "USER": "root", "HOME": "/root", "MAIL": "/var/mail/root", "SSH_CONNECTION": "192.168.51.100 36208 192.168.46.3 22", "XDG_SESSION_ID": "3", "XDG_SESSION_CLASS": "user", "_": "/bin/sh"}, "ansible_effective_group_id": 0, "ansible_userspace_bits": "64", "ansible_ssh_host_key_ecdsa_public": "AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBMTGUP+E0z7uHyF4F+MWV3vov4nw/Ypjq25IQtvH+ZKvj2+jEfYutGJ/Ypr9xxXebh+h7BBW+sxqXfKyylk6M2Y=", "ansible_real_user_id": 0, "ansible_distribution_release": "buster", "ansible_ssh_host_key_dsa_public_keytype": "ssh-dss", "ansible_apparmor": {"status": "enabled"}, "ansible_cmdline": {"console": "ttyS0,115200", "earlyprintk": "ttyS0,115200", "root": "UUID=b5850e77-bb82-4674-ab82-8e8dccda3f0e", "net.ifnames": "0", "consoleblank": "0", "biosdevname": "0", "BOOT_IMAGE": "/boot/vmlinuz-4.19.0-9-cloud-amd64", "ro": true, "systemd.show_status": "true"}, "ansible_ssh_host_key_ed25519_public_keytype": "ssh-ed25519", "ansible_selinux": {"status": "Missing selinux Python library"}, "ansible_ssh_host_key_ecdsa_public_keytype": "ecdsa-sha2-nistp256", "ansible_os_family": "Debian", "ansible_userspace_architecture": "x86_64", "ansible_pkg_mgr": "apt", "ansible_user_shell": "/bin/bash", "ansible_user_uid": 0, "ansible_distribution": "Debian", "ansible_user_dir": "/root", "ansible_dns": {"nameservers": ["127.0.0.1"]}, "ansible_ssh_host_key_dsa_public": "AAAAB3NzaC1kc3MAAACBAIop5fGYOCqo+968Zkcw/pgizNP7HXaIcZLS9x5cJfhOqtZZO1S0gJlxHMXled22K1ZIKZVd7QruZuS3p8GzWem2u6a4dzZZ85n6sNm3rJZFNekYNsXwctduE59qbL0D5++7NYfeaCsSpWHi8e+7Ye2k4kvbwyRyYXBytzJtlgzpAAAAFQDAVnB/mR9Zp8IKVptTjjqaayCR3QAAAIA8qkrFfaK5aKk7Z9Qv9VkLEYPfTV2nzTLmFV3ynQe92sh0Z/qCAkW7dhY3glrq9iqIK8NXgx1CNIcIiyT5IbpdUgbRLW70ehKYdkxhLyGyC07fyKY/UhPQRw60Fv4WVyZfePp6fNxOQpWku2LNqILrGSyMFqBQAO+d9oIB6y2zjgAAAIBLYOz/mO149c+mnTi0hsOOf/FtVCTtc98NEV0lQxUXsQt+4pPg4LOWh+hjDNabIuk/jfXqJfxYEBk5jxYd4W/BxunFsTOLBrPjG1b7V/jq8pRdRzqverjkmfpSHxxGKyMQmiPyaOXomxhreXPWwysrSUm1qnYHaCxdTBb/4qP0AA==", "ansible_selinux_python_present": false, "ansible_effective_user_id": 0, "ansible_hostname": "debiantest", "ansible_real_group_id": 0, "ansible_lsb": {"release": "10", "major_release": "10", "codename": "buster", "id": "Debian", "description": "Debian GNU/Linux 10 (buster)"}, "ansible_proc_cmdline": {"console": ["tty0", "ttyS0,115200"], "earlyprintk": "ttyS0,115200", "root": "UUID=b5850e77-bb82-4674-ab82-8e8dccda3f0e", "net.ifnames": "0", "consoleblank": "0", "biosdevname": "0", "BOOT_IMAGE": "/boot/vmlinuz-4.19.0-9-cloud-amd64", "ro": true, "systemd.show_status": "true"}, "ansible_local": {}, "ansible_machine_id": "f8d97d23945344088763643c62143ab8", "ansible_ssh_host_key_rsa_public_keytype": "ssh-rsa", "ansible_system_capabilities_enforced": "True", "ansible_user_gecos": "root", "ansible_system_capabilities": ["cap_chown", "cap_dac_override", "cap_dac_read_search", "cap_fowner", "cap_fsetid", "cap_kill", "cap_setgid", "cap_setuid", "cap_setpcap", "cap_linux_immutable", "cap_net_bind_service", "cap_net_broadcast", "cap_net_admin", "cap_net_raw", "cap_ipc_lock", "cap_ipc_owner", "cap_sys_module", "cap_sys_rawio", "cap_sys_chroot", "cap_sys_ptrace", "cap_sys_pacct", "cap_sys_admin", "cap_sys_boot", "cap_sys_nice", "cap_sys_resource", "cap_sys_time", "cap_sys_tty_config", "cap_mknod", "cap_lease", "cap_audit_write", "cap_audit_control", "cap_setfcap", "cap_mac_override", "cap_mac_admin", "cap_syslog", "cap_wake_alarm", "cap_block_suspend", "cap_audit_read+ep"], "ansible_python": {"executable": "/usr/bin/python", "version": {"micro": 16, "major": 2, "releaselevel": "final", "serial": 0, "minor": 7}, "type": "CPython", "has_sslcontext": true, "version_info": [2, 7, 16, "final", 0]}, "ansible_fqdn": "debiantest", "ansible_user_gid": 0, "ansible_nodename": "debiantest", "ansible_system": "Linux", "ansible_machine": "x86_64", "ansible_distribution_major_version": "10", "ansible_kernel": "4.19.0-9-cloud-amd64", "ansible_python_version": "2.7.16"}}\r\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.46.3 closed.\r\n')
Including module_utils file ansible/__init__.py
Including module_utils file ansible/module_utils/__init__.py
Including module_utils file ansible/module_utils/basic.py
Including module_utils file ansible/module_utils/_text.py
Including module_utils file ansible/module_utils/common/_collections_compat.py
Including module_utils file ansible/module_utils/common/__init__.py
Including module_utils file ansible/module_utils/common/_json_compat.py
Including module_utils file ansible/module_utils/common/_utils.py
Including module_utils file ansible/module_utils/common/file.py
Including module_utils file ansible/module_utils/common/parameters.py
Including module_utils file ansible/module_utils/common/collections.py
Including module_utils file ansible/module_utils/common/process.py
Including module_utils file ansible/module_utils/common/sys_info.py
Including module_utils file ansible/module_utils/common/text/converters.py
Including module_utils file ansible/module_utils/common/text/__init__.py
Including module_utils file ansible/module_utils/common/text/formatters.py
Including module_utils file ansible/module_utils/common/validation.py
Including module_utils file ansible/module_utils/common/warnings.py
Including module_utils file ansible/module_utils/compat/selectors.py
Including module_utils file ansible/module_utils/compat/__init__.py
Including module_utils file ansible/module_utils/compat/_selectors2.py
Including module_utils file ansible/module_utils/distro/__init__.py
Including module_utils file ansible/module_utils/distro/_distro.py
Including module_utils file ansible/module_utils/parsing/convert_bool.py
Including module_utils file ansible/module_utils/parsing/__init__.py
Including module_utils file ansible/module_utils/pycompat24.py
Including module_utils file ansible/module_utils/six/__init__.py
Using module file /home/user/github/ansible/lib/ansible/modules/find.py
<192.168.46.3> PUT /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmpa1dl5mj3 TO /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_find.py
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: PlayContext set sftp_extra_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 '[192.168.46.3]'
<192.168.46.3> (0, b'sftp> put /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmpa1dl5mj3 /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_find.py\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug2: Remote version: 3\r\ndebug2: Server supports extension "posix-rename@openssh.com" revision 1\r\ndebug2: Server supports extension "statvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "fstatvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "hardlink@openssh.com" revision 1\r\ndebug2: Server supports extension "fsync@openssh.com" revision 1\r\ndebug3: Sent message fd 3 T:16 I:1\r\ndebug3: SSH_FXP_REALPATH . -> /root size 0\r\ndebug3: Looking up /home/user/.ansible/tmp/ansible-local-132512hmn7kbq/tmpa1dl5mj3\r\ndebug3: Sent message fd 3 T:17 I:2\r\ndebug3: Received stat reply T:101 I:2\r\ndebug1: Couldn\'t stat remote file: No such file or directory\r\ndebug3: Sent message SSH2_FXP_OPEN I:3 P:/root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_find.py\r\ndebug3: Sent message SSH2_FXP_WRITE I:4 O:0 S:32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 4 32768 bytes at 0\r\ndebug3: Sent message SSH2_FXP_WRITE I:5 O:32768 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:6 O:65536 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:7 O:98304 S:21382\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 5 32768 bytes at 32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 6 32768 bytes at 65536\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 7 21382 bytes at 98304\r\ndebug3: Sent message SSH2_FXP_CLOSE I:4\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 192.168.46.3 '/bin/sh -c '"'"'chmod u+x /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/ /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_find.py && sleep 0'"'"''
<192.168.46.3> (0, b'', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 -tt 192.168.46.3 '/bin/sh -c '"'"'/usr/bin/python /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/AnsiballZ_find.py && sleep 0'"'"''
<192.168.46.3> (0, b'\r\n{"files": [{"uid": 0, "woth": true, "mtime": 1588006977.0, "inode": 4722, "isgid": false, "size": 14, "roth": true, "isuid": false, "isreg": false, "pw_name": "root", "gid": 0, "ischr": false, "wusr": true, "xoth": true, "rusr": true, "nlink": 1, "issock": false, "rgrp": true, "gr_name": "root", "path": "/sbin/shutdown", "xusr": true, "atime": 1596111913.18, "isdir": false, "ctime": 1591791848.7132318, "isblk": false, "xgrp": true, "dev": 2049, "wgrp": true, "isfifo": false, "mode": "0777", "islnk": true}, {"uid": 0, "woth": true, "mtime": 1588006977.0, "inode": 4722, "isgid": false, "size": 14, "roth": true, "isuid": false, "isreg": false, "pw_name": "root", "gid": 0, "ischr": false, "wusr": true, "xoth": true, "rusr": true, "nlink": 1, "issock": false, "rgrp": true, "gr_name": "root", "path": "/usr/sbin/shutdown", "xusr": true, "atime": 1596111913.18, "isdir": false, "ctime": 1591791848.7132318, "isblk": false, "xgrp": true, "dev": 2049, "wgrp": true, "isfifo": false, "mode": "0777", "islnk": true}], "changed": false, "msg": "", "exusered": 500, "invocation": {"module_args": {"excludes": null, "paths": ["/sbin", "/usr/sbin", "/usr/local/sbin"], "file_type": "any", "age": null, "contains": null, "recurse": false, "age_stamp": "mtime", "patterns": ["shutdown"], "depth": null, "get_checksum": false, "use_regex": false, "follow": false, "hidden": false, "size": null}}, "matched": 2}\r\n', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.46.3 closed.\r\n')
community.general.shutdown: shutting down server...
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 -tt 192.168.46.3 '/bin/sh -c '"'"'/sbin/shutdown -h 0 "Shut down initiated by Ansible" && sleep 0'"'"''
<192.168.46.3> (0, b"Shutdown scheduled for Thu 2020-07-30 15:23:24 UTC, use 'shutdown -c' to cancel.\r\n", b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.46.3 closed.\r\n')
<192.168.46.3> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.46.3> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<192.168.46.3> SSH: ansible_password/ansible_ssh_password not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.46.3> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User="root")
<192.168.46.3> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.46.3> SSH: PlayContext set ssh_common_args: ()
<192.168.46.3> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/user/.ansible/cp/a7e952dc64)
<192.168.46.3> SSH: EXEC /usr/bin/ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="root"' -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/a7e952dc64 192.168.46.3 '/bin/sh -c '"'"'rm -f -r /root/.ansible/tmp/ansible-tmp-1596122604.463402-13282-199918880950791/ > /dev/null 2>&1 && sleep 0'"'"''
<192.168.46.3> (255, b'', b'OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly\r\n')
fatal: [192.168.46.3]: FAILED! => {
    "msg": "Failed to connect to the host via ssh: OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020\r\ndebug1: Reading configuration data /home/user/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug2: resolve_canonicalize: hostname 192.168.46.3 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13266\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly"
}
```

Andersson007 commented:
i haven't dig into the code, maybe to write something instead of `validate`, pass the `module` object to this and run `module.exit_json()` there.

aminvakil commented:
The problem resists in ansible trying to remove tmp files after executing the module, I can verify this by setting `export ANSIBLE_KEEP_REMOTE_FILES=1`, but I cannot find a way to set it in `shutdown.py`.

Andersson007 commented:
cc @mattclay @samdoran ^^

aminvakil commented:
Could anyone please check copyrights and authors too?

Andersson007 commented:
@aminvakil usually a very efficient way is to ask folks in #ansible-devel IRC channel after 3 pm UTC because they are in the US (if nobody replies, ask sdoran: and mattclay: directly)

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+
+        distribution = self.get_distribution(task_vars)
+
+        # Get the original connection_timeout option var so it can be reset after
+        original_connection_timeout = None
+        try:
+            original_connection_timeout = self._connection.get_option('connection_timeout')
+            display.debug("{action}: saving original connect_timeout of {timeout}".format(action=self._task.action, timeout=original_connection_timeout))
+        except KeyError:
+            display.debug("{action}: connect_timeout connection option has not been set".format(action=self._task.action))
```
```suggestion
```
I don't think this is needed anymore.

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/83b05b476ab4930c1054ed7bb4126bf818a4e0f3.

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,73 @@
+      - On Solaris and FreeBSD, this will be seconds.
+    type: int
+    default: 0
+  connect_timeout:
+    description:
+      - Maximum seconds to wait for a successful connection to the managed hosts before trying again.
+      - If unspecified, the default setting for the underlying connection plugin is used.
+    type: int
```
```suggestion
```
Same here. This was used in `reboot` for the reconnect, if I understand it correctly.

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/83b05b476ab4930c1054ed7bb4126bf818a4e0f3.

felixfontein commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,73 @@
+
+RETURN = r'''
+shutdown:
+  description: true if the machine has shut down
```
```suggestion
  description: C(true) if the machine has been shut down.
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/a11f7192301e45b2d90dd3afa137df8f03e6a4c7.

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+
+    def _check_delay(self, key, default):
+        """Ensure that the value is positive or zero"""
+        value = int(self._task.args.get(key, self._task.args.get(key + '_sec', default)))
```
Why `key + '_sec'`?

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/879ce8345257af1c1c25cd46adbfe7b95f2565b6.
Actually I don't even why was the reason for second `self._task.args.get(key, default)`, I checked and they're all the same, therefore I have removed the second `args.get` completely.
I think this is sufficient:
`value = int(self._task.args.get(key, default))`

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+    def get_shutdown_command_args(self, distribution):
+        args = self._get_value_from_facts('SHUTDOWN_COMMAND_ARGS', distribution, 'DEFAULT_SHUTDOWN_COMMAND_ARGS')
+        # Convert seconds to minutes. If less that 60, set it to 0.
+        delay_min = self.delay // 60
```
How about storing the value of `self.delay` as `delay_sec`, so you don't have to call `self.delay` a second time in this function?

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/d275da6a06707613c0cef75e8e6cac0b7b7173a2.

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+    DEFAULT_SHUTDOWN_COMMAND_ARGS = '-h {delay_min} "{message}"'
+    DEFAULT_SUDOABLE = True
+
+    DEPRECATED_ARGS = {}
```
```suggestion
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/a156e49f426643ea8235136b909a9180cafa8ac6.

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+        self._shutdown_command = full_path[0]
+        return self._shutdown_command
+
+    def deprecated_args(self):
+        for arg, version in self.DEPRECATED_ARGS.items():
+            if self._task.args.get(arg) is not None:
+                display.warning("Since Ansible {version}, {arg} is no longer a valid option for {action}".format(
+                    version=version,
+                    arg=arg,
+                    action=self._task.action))
```
```suggestion
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/a156e49f426643ea8235136b909a9180cafa8ac6.

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+        if task_vars is None:
+            task_vars = {}
+
+        self.deprecated_args()
```
```suggestion
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/a156e49f426643ea8235136b909a9180cafa8ac6.

felixfontein commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,234 @@
+            msg = 'Running {0} with local connection would shutdown the control node.'.format(self._task.action)
+            return {'changed': False, 'elapsed': 0, 'shutdown': False, 'failed': True, 'msg': msg}
+
+        if self._play_context.check_mode:
```
How about running everything when `self._play_context.check_mode` is `True` except the `shutdown_result = self._low_level_execute_command(shutdown_command_exec, sudoable=self.DEFAULT_SUDOABLE)` in the previous function? That way, check mode is even more realistic.

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/e78ae50651498ad4efb7525dfcd4c49cf40bb467.

Andersson007 commented:
@aminvakil if it's 100% ready for review, i could take a look on Saturday (but not absolutely sure, i have pretty tough work schedule these days). thanks for doing this!

aminvakil commented:
@Andersson007 Sorry to bother you multiple times to review this PR, every time I request a review in my opinion it's 100% ready for review, but there has been times which it had bugs, defects, extra stuff, etc., accept my apologies for those. Please review this anytime which is appropriate for you, thanks again!

Andersson007 commented:
@aminvakil ah, no problem, you're always welcome!:) no need to apologise, i understand. i often do the same;)

Andersson007 commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,214 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import random
+import time
+
+from datetime import datetime, timedelta
```
```suggestion
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/21bf25b0264a24e0cf5e2dc03f4c68a5b13d9401.

Andersson007 commented on plugins/action/system/shutdown.py:
```diff
@@ -0,0 +1,214 @@
+from ansible.errors import AnsibleError, AnsibleConnectionFailure
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.common.collections import is_string
+from ansible.module_utils.common.validation import check_type_str
```
```suggestion
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/21bf25b0264a24e0cf5e2dc03f4c68a5b13d9401.

Andersson007 commented on plugins/modules/system/shutdown.py:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2018, Ansible Project
```
```suggestion
# Copyright: (c) 2020, Ansible Project
```

aminvakil replied:
Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/21bf25b0264a24e0cf5e2dc03f4c68a5b13d9401.

aminvakil replied:
Will mark this as resolved after double-checking with `sdoran` on IRC.

Andersson007 commented:
@aminvakil  if you tried this manually on several distributions and it works as expected everywhere, so LGTM
Could you also join IRC and ask `mattclay` and `sdoran` in #ansible-devel IRC channel after 3 pm UTC about copyright headers and code maybe too.

Andersson007 commented:
@felixfontein do you have anything to add?

Andersson007 commented:
@aminvakil also feel free to mark the fixed comments as resolved

aminvakil commented:
@Andersson007 I've checked on centos 7,8 and debian 10. If an integration test could be tested on it that would be awesome, maybe if somehow it can be executed after all other tests(?).

About copyright issue I have asked in IRC and `sdoran` told me that there is nothing needed to change, but I've changed it as you've said, will double check this on ansible-devel IRC at 3pm UTC.

felixfontein commented:
@Andersson007 no, I think it's ready.
@aminvakil adding integration tests won't be possible without either creating new VMs (that could be shut down), or changing ansible-test (which I guess is not worth the effort).

aminvakil commented:
@felixfontein What about creating a shutdown task with a delay, check if a shutdown has been scheduled, and then cancel the shutdown.
I assume this can be done on linux at least.

felixfontein commented:
The only thing you could do is also return the executed command, and then add integration tests with `check_mode: true` which check that the correct command was used for the platforms involved.

aminvakil commented:
ready_for_review

Andersson007 commented:
why have you removed CI tests?

aminvakil commented:
> why have you removed CI tests?

There was a discussion on `ansible-devel` and some core developers thought that running it with `check_mode` doesn't help on testing the module, I have sent you an email containing the chats.

felixfontein commented:
Please re-add the CI tests.

aminvakil commented:
/rebuild_failed

felixfontein commented on tests/integration/targets/shutdown/tasks/main.yml:
```diff
@@ -0,0 +1,36 @@
+---
+- block:
+    - name: execute shutdown with check_mode true
+      community.general.shutdown:
```
How about adding a few more tests, also with `delay` and `msg`?

aminvakil commented:
How can a specific linux distribution version gets skipped on aliases?
I've tried before as https://docs.ansible.com/ansible/latest/dev_guide/testing/sanity/integration-aliases.html#skipping said in this commit: https://github.com/ansible-collections/community.general/pull/700/commits/7b40628bf229bf1bb1cf04714a39919f2775cae6
But it didn't work, what am I missing?

In ubuntu 18 systemd-sysv package which provides /sbin/shutdown, /sbin/reboot, /sbin/halt, /sbin/poweroff is not a dependency of systemd. Therefore systemd can be installed without these symlinks.

Andersson007 commented:
```
skip/osx
```
in tests/integration/targets/postgresql_set/aliases file, for example, worked for me

felixfontein commented:
`skip/` only works for some platforms. In general, it's best to do a `block:` with `when:`.

For Ubuntu 18, you could also simply install `systemd-sysv` at the beginning of the tests.

felixfontein commented:
(Or actually support `systemctl halt` / `systemctl poweroff` ;) )

aminvakil commented:
> ```
> skip/osx
> ```
> 
> in tests/integration/targets/postgresql_set/aliases file, for example, worked for me

> `skip/` only works for some platforms. In general, it's best to do a `block:` with `when:`.
> 
> For Ubuntu 18, you could also simply install `systemd-sysv` at the beginning of the tests.

OK then, at first install `systemd-sysv` on ubuntu 18 and then the test begins.

> (Or actually support `systemctl halt` / `systemctl poweroff` ;) )

I think it's not necessary for now, maybe a PR later:)

Andersson007 commented:
yep, `when: ansible_os_distribution != 'Ubuntu'` should also work

aminvakil commented:
> yep, `when: ansible_os_distribution != 'Ubuntu'` should also work

I've done that before in https://github.com/ansible-collections/community.general/pull/700/commits/b772e2facca54397bbba3afef8f4764438a7f8de.

But as I've changed the structure of test now, I should either put the whole test in a block and run it when it's not ubuntu 18, or install systemd-sysv when it's ubuntu 18 before anything else and remove it after the test is complete.

I think the latter one also tests ubuntu 18 and it's a better approach, isn't it?

felixfontein commented:
> > (Or actually support `systemctl halt` / `systemctl poweroff` ;) )
> 
> I think it's not necessary for now, maybe a PR later:)

That would also be more complicated, since it would have to detect systemd.

> I think the latter one also tests ubuntu 18 and it's a better approach, isn't it?

:+1:

felixfontein commented on tests/integration/targets/shutdown/tasks/main.yml:
```diff
@@ -69,3 +76,9 @@
+  apt:
+    name: systemd-sysv
+    state: absent
+  when: systemd_sysv_install is changed and ansible_distribution == 'Ubuntu' and ansible_distribution_major_version is version('18', '>=')
```
```suggestion
  when: systemd_sysv_install is changed
```
This should also work (since skipped => not changed), and make it easier to change the conditions when this package should be installed (you don't have to adjust the condition here as well).

aminvakil replied:
Sure. Fixed in https://github.com/ansible-collections/community.general/pull/700/commits/9692bb42680e74ae795fd275a8feaf956c354948.

felixfontein commented:
Failing CI is unrelated.

aminvakil commented:
> or install systemd-sysv when it's ubuntu 18 before anything else and remove it after the test is complete.

I don't know whether this a problem or not, but `systemd-sysv` installs and upgrades additional packages as well and they won't be removed and downgraded when `systemd-sysv` gets removed.
```
The following additional packages will be installed:
 libnss-systemd libsystemd0 systemd
The following packages will be upgraded:
libsystemd0 systemd
```
Therefore the test for this module doesn't leave the container as it was before the test.
Should  I revert the commit and ignore ubuntu 18 then or leave it as it is now?

felixfontein commented:
@aminvakil that's a side-effect of package installing... just keep it, it shouldn't disturb tests.

Andersson007 commented:
can we merge this?

felixfontein commented:
Yes we can!

@aminvakil thanks for contributing this!
@Andersson007 thanks for reviewing!

Andersson007 commented:
@aminvakil thanks for the new module!
@felixfontein thanks for reviewing!

aminvakil commented:
Please accept my apologies for haven't looking it carefully, but osx test doesn't get checked.

https://github.com/ansible-collections/community.general/commit/c475effeedbf643de5d2967eee001b496d52a033#diff-3bbcd8dc5bda13e460f5ceed5609938aR57

aminvakil commented:
@felixfontein @Andersson007 
I guess I'll break some records (not good records) if I just create a PR fixing the new module:)
But let me know if I should.

felixfontein commented:
@aminvakil it's totally ok to create another PR for this ;)

felixfontein commented:
#769

aminvakil commented:
@felixfontein @Andersson007 Thanks for reviewing and helping on this module.
